src/Parsing/ParserState.php

```diff
@@ -30,7 +30,7 @@ class ParserState
     /**
      * @var array<int, string>
      */
-    private $characters;
+    private $characters = [];
```

Comment:
This change is not needed (as we set it in `setCharset()`, which is called from the constructor. Also, it's misleading as this property does not have a default value after construction.
```suggestion
    private $characters;
```

Reply:
This change gives it a default value, ensuring it is always an array, in case the construction chain somehow fails.

Reply:
AFAIK, if construction fails (e.g., if an exception is thrown during construction), there will be no constructed object: https://onlinephp.io/c/7a627

(Also, I think if construction fails, we shouldn't try return a half-broken object, but have properly not-yet-initialized properties if they have not been initialized with meaningful data.)

Reply:
Would PHPStan be happy that it's always an array, even if only initialized to one by a method called by the constructor?

Reply:
Yes, AFAIK.